### ISSUE TITLE
Get negative value suing keras 'kld' loss

### ISSUE THREAD

**Issue by HuangBo-Terraloupe** [keras-team/keras#9754]
Hello everyone,

As we all know, the kld loss can not be negative, I am training a regression model, and get negative values.
Here is my model: 
# model
base_model = VGG16(input_shape=(360, 480, 3), weights='imagenet', include_top=False)
x = base_model.layers[-2].output
x = MaxPool2D(pool_size=(2, 2), padding='same', strides=(1, 1), name='block5_pool')(x)
x = Conv2D(32, (7, 7), activation='relu', padding='same', name='block5_conv5')(x)
x = Conv2D(8, (7, 7), activation='relu', padding='same', name='block5_conv6')(x)
x = Conv2D(1, (7, 7), activation='relu', padding='same', name='block5_conv7')(x)
x = Flatten(name='flatten')(x)
prediction = Activation('softmax')(x)  # problem come in!!!!!!!!!!!
model = Model(inputs=base_model.input, outputs=prediction)

# compile
adam = Adam(lr=1e-5, beta_1=0.9, beta_2=0.999, epsilon=1e-8, decay=0.0)
model.compile(optimizer=adam, loss='kld', metrics=['accuracy'])

The problem is, if I add a softmax layer at end of the model, the loss is positive, which is fine, but the loss is around 32, it is really big. But if I remove the softmax layer, the loss becomes negative. 

For the input and output, input are images, I normalize the images to 0-1, and labels also 0-1. 
My point is, this is a regression model, I do not want to add a softmax layer at end of the model, but the loss becomes negative, which is not right. Is there someone has a idea? How to solve the problem?

**Comment by muyinanhai** [comment]
kld loss is negtive when y_true is small than y_pred
```python
def kullback_leibler_divergence(y_true, y_pred):
    y_true = K.clip(y_true, K.epsilon(), 1)
    y_pred = K.clip(y_pred, K.epsilon(), 1)
    return K.sum(y_true * K.log(y_true / y_pred), axis=-1)
```


BTW, kld loss is not the same as kld, you can have look at [kullback-leibler-divergence-negative-values](https://stats.stackexchange.com/questions/41297/kullback-leibler-divergence-negative-values)

**Comment by HuangBo-Terraloupe** [comment]
@muyinanhai 
You mean this keras 'kld' loss is not the real kld loss? But the example you show me seems only calculate one instance in the equation, but here keras use 'sum'.
I found a tensorflow implementation of kld loss:

def kl_divergence(labels, prediction, epsilon=1e-7):

    # normalize to distribution
    prediction /= (tf.reduce_sum(prediction, axis=1, keep_dims=True) + epsilon)
    labels /= (tf.reduce_sum(labels, axis=1, keep_dims=True) + epsilon)

    result = tf.reduce_mean(tf.reduce_sum(labels * tf.log((labels / (prediction + epsilon)) + epsilon), axis=1))
    return result


#test code:
gt = np.random.rand(1, 10)
print('the ground truth is:', gt)
gt = tf.convert_to_tensor(gt)

# pre = np.random.rand(1, 10)
pre = np.zeros([1, 10])
print('the prediction is:', pre)
pre = tf.convert_to_tensor(pre)

sess = tf.InteractiveSession()
print(kl_divergence(gt, pre).eval())
It seems right? But when the gt and pre are same, I also get negative value: -8.999988943532526e-07.
Maybe is the reason of epsilon

**Comment by jimmylevell** [comment]
Could somebody elaborate why the keras KL divergence implementation can create negative values? And how this suits the definition of the KL divergence?

And what preciously is meant with the following statement? 
> BTW, kld loss is not the same as kld, you can have look at kullback-leibler-divergence-negative-values
